astropy/cosmology/tests/test_core.py

```diff
@@ -0,0 +1,325 @@
+
+            def __init__(self, param, *, name=None, meta=None):
+                cls = self.__class__
+                with u.add_enabled_equivalencies(cls.param.equivalencies):
+                    self._param = param << cls.param.unit
+
+        assert Example(1).param == 1 * u.eV
+        assert Example(1 * u.eV).param == 1 * u.eV
+        assert Example(1 * u.J).param == (1 * u.J).to(u.eV)
+        assert Example(1 * u.kg).param == (1 * u.kg).to(u.eV, u.mass_energy())
+
+
+class TestCosmology(ParameterTestMixin, metaclass=abc.ABCMeta):
+    """Test :class:`astropy.cosmology.Cosmology`"""
+
+    def setup_class(self):
+        """
+        Setup for testing.
+        Cosmology should not be instantiated, so tests are done on a subclass.
+        """
+        class SubCosmology(Cosmology):
+            pass
+
+        self.cls = SubCosmology
+        self.cls_args = (70 * (u.km / u.s / u.Mpc),)  # nothing real
+        self.cls_kwargs = dict(name="test", meta={"a": "b"})
+
+    def teardown_class(self):
+        _COSMOLOGY_CLASSES.pop("TestCosmology.setup_class.<locals>.SubCosmology", None)
+
+    @pytest.fixture
+    def cosmo_cls(self):
+        return self.cls
+
+    @pytest.fixture
+    def cosmo(self):
+        """The cosmology instance with which to test."""
+        return self.cls(*self.cls_args, **self.cls_kwargs)
+
+    # ===============================================================
+    # Method & Attribute Tests
+
+    def _cosmo_test_init_attr(self, cosmo):
+        """Helper function for testing ``__init__``"""
+        assert hasattr(cosmo, "_name")
+        assert cosmo._name is None or isinstance(cosmo._name, str)
+
+        assert hasattr(cosmo, "meta")
+        assert isinstance(cosmo.meta, dict)
+
+    def test_init(self, cosmo_cls):
+        # Cosmology accepts any args, kwargs
+        cosmo1 = cosmo_cls(1, 2, 3, 4, 5, a=1, b=2, c=3, d=4, e=5)
+        self._cosmo_test_init_attr(cosmo1)
+
+        # but only "name" and "meta" are used
+        cosmo2 = cosmo_cls(name="test", meta={"m": 1})
+        self._cosmo_test_init_attr(cosmo2)
+        assert cosmo2.name == "test"
+        assert cosmo2.meta["m"] == 1
+
+    # ---------------------------------------------------------------
+
+    def test_is_equivalent(self, cosmo):
+        """
+        Test :meth:`astropy.cosmology.Cosmology.is_equivalent`.
+        """
+        # to self
+        assert cosmo.is_equivalent(cosmo)
+
+        # same class, different instance
+        newclone = cosmo.clone(name="cloned")
+        assert cosmo.is_equivalent(newclone)
+        assert newclone.is_equivalent(cosmo)
+
+        # different class
+        assert not cosmo.is_equivalent(2)
+
+
+class CosmologySubclassTest(TestCosmology):
+    """
+    Test subclasses of :class:`astropy.cosmology.Cosmology`.
+    This is broken away from ``TestCosmology``, because |Cosmology| is/will be
+    an ABC and subclasses must override some methods.
+    """
+
+    @abc.abstractmethod
```

Comment:
Maybe because of this. But is it actually used?

Reply:
Yeah, that's one reason. ``FLRWSubclassTest`` is used here for all the FLRW-subclass tests, like ``TestLambdaCDM``.
And if someone wanted to write their own FLRW subclass and use the test suite, then they too would have to define a ``setup_class()``.

Reply:
When I start migrating the tests from ``test_cosmology`` to the new test suite I'll probably add 

```python
@abc.abstractmethod
def test_w(self, cosmo):
    """`FLRW.w` is abstract and subclasses must provide a test."""
```